components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt

```diff
@@ -331,19 +334,28 @@ open class GleanInternalAPI internal constructor () {
             return false
         }
 
-        var hasPingContent = false
+        val pingSerializationTasks: MutableList<Job> = mutableListOf()
         for (pingName in pingNames) {
-            val hasContent = assembleAndSerializePing(pingName)
-            hasPingContent = hasPingContent || hasContent
+            assembleAndSerializePing(pingName)?.let {
+                pingSerializationTasks.add(it)
+            }
         }
 
         // It should only take a single PingUploadWorker to process all queued pings, so we only
         // want to have one scheduled at a time.
-        if (hasPingContent) {
-            PingUploadWorker.enqueueWorker()
+        if (pingSerializationTasks.any()) {
+            // Launch the serialization tasks in a background thread as they run blocking and we
+            // don't necessarily want to block any calling thread.
+            Dispatchers.API.launch {
+                runBlocking {
```

Comment:
You are right about runBlocking not necessary here, it was necessary before I used a coroutine to launch this in the background as `.joinAll()` requires a separate coroutine scope to run in.  When I moved this to run on the glean coroutine scope, that removed the need for runBlocking.

If I were to use `invokeOnCompletion` I would have to chain them together, running each additional serialization job in the next `invokeOnCompletion`  the code was uglier than collecting all of the jobs and waiting on them to join before scheduling the Worker.

Reply:
> If I were to use `invokeOnCompletion` I would have to chain them together, running each additional serialization job in the next `invokeOnCompletion` the code was uglier than collecting all of the jobs and waiting on them to join before scheduling the Worker.

Why would you? I don't think the code above would produce a result that's any different than the current implementation

Reply:
It waits on all the pings to be written with the joinAll() before enqueueing the worker..  that's what the bug is about...   I don't understand why you think this would execute the same.  If I have a list of `Job`s that each represent writing a ping to disk, and I wait for all of those jobs to join before enqueuing the `PingUploadWorker`, it should guarantee that everything is written to disk before trying to upload them...

Reply:
Right now, it loops through spawning the Job's but doesn't wait on any of them before enqueueing the Worker, so IO may still be going on when the Worker runs...

Reply:
What I mean is that

```
Dispatchers.API.launch {
  pingSerializationTasks.joinAll()
}.invokeOnCompletion { PingUploadWorker.enqueueWorker() }
```

should be equivalent to

```
Dispatchers.API.launch {
  pingSerializationTasks.joinAll()
  PingUploadWorker.enqueueWorker()
}
```

Reply:
I see what you are saying now...    it looks more Kotlin'ish, so I'll bite

Reply:
Sorry, I realized too late that I was unclear :)

Reply:
Can't do it this way...   because of how we subvert `Dispatchers.API.launch` to run synchronously for tests, it doesn't always return a `Job` to `invokeOnCompletion()` on (such as any time glean is in test mode), so according to the [docs](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/join-all.html) for `joinAll()`, it still blocks the current coroutine scope until all jobs are complete so it shouldn't execute `PingUploadWorker.enqueueWorker()` until the `joinAll()` is complete.

So, while this way would work great for release, it breaks our tests so I am going to keep it like I had it.